SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -0,0 +1,338 @@
+        [YapDatabaseViewGrouping withObjectBlock:^NSString *_Nullable(YapDatabaseReadTransaction *_Nonnull transaction,
+            NSString *_Nonnull collection,
+            NSString *_Nonnull key,
+            id _Nonnull object) {
+            if (![object isKindOfClass:[OWSMessageProcessingJob class]]) {
+                OWSFail(@"Unexpected object: %@ in collection: %@", object, collection) return nil;
+            }
+
+            // Arbitrary string - all in the same group. We're only using the view for sorting.
+            return OWSMessageProcessingJobFinderExtensionGroup;
+        }];
+
+    YapDatabaseViewOptions *options = [YapDatabaseViewOptions new];
+    options.allowedCollections =
+        [[YapWhitelistBlacklist alloc] initWithWhitelist:[NSSet setWithObject:[OWSMessageProcessingJob collection]]];
+
+    return [[YapDatabaseView alloc] initWithGrouping:grouping sorting:sorting versionTag:@"1" options:options];
+}
+
+
++ (void)syncRegisterDatabaseExtension:(YapDatabase *)database
+{
+    YapDatabaseView *existingView = [database registeredExtension:OWSMessageProcessingJobFinderExtensionName];
+    if (existingView) {
+        // already initialized
+        return;
+    }
+    [database registerExtension:[self databaseExension] withName:OWSMessageProcessingJobFinderExtensionName];
+}
+
+@end
+
+#pragma mark - Queue Processing
+
+@interface OWSMessageProcessingQueue : NSObject
+
+@property (nonatomic, readonly) TSMessagesManager *messagesManager;
+@property (nonatomic, readonly) OWSMessageProcessingJobFinder *finder;
+
+- (instancetype)initWithMessagesManager:(TSMessagesManager *)messagesManager
+                                 finder:(OWSMessageProcessingJobFinder *)finder NS_DESIGNATED_INITIALIZER;
+- (instancetype)init NS_UNAVAILABLE;
+
+@end
+
+@implementation OWSMessageProcessingQueue
+
+- (instancetype)initWithMessagesManager:(TSMessagesManager *)messagesManager
+                                 finder:(OWSMessageProcessingJobFinder *)finder
+{
+    OWSSingletonAssert();
+
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    _messagesManager = messagesManager;
+    _finder = finder;
+
+    return self;
+}
+
+#pragma mark - instance methods
+
+- (void)enqueueEnvelopeForProcessing:(OWSSignalServiceProtosEnvelope *)envelope
+{
+    [self.finder addJobForEnvelope:envelope];
+}
+
+- (void)drainQueue
+{
+    dispatch_async(self.class.serialGCDQueue, ^{
+        OWSMessageProcessingJob *_Nullable job = [self.finder nextJob];
+        if (job == nil) {
+            DDLogVerbose(@"%@ Queue is drained", self.tag);
+            return;
+        }
+
+        [self processJob:job
+              completion:^{
+                  [self drainQueue];
+              }];
+    });
+}
+
+- (void)processJob:(OWSMessageProcessingJob *)job completion:(void (^)())completion
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self.messagesManager processEnvelope:job.envelopeProto
```

Comment:
So if we were going to build upon this work to retain jobs that couldn't be processed (e.g. messages for unknown groups), I assume this completion would have a success parameter, and we'd add a "failed" flag and a "thread id" to jobs and we'd modify the view to ignore failed jobs, and when we got a group update we'd clear the failed flag for all jobs in that thread?